src/invidious/videos.cr

```diff
@@ -871,30 +898,61 @@ def extract_video_info(video_id : String, proxy_region : String? = nil, context_
     params[f] = player_response[f] if player_response[f]?
   end
 
-  params["relatedVideos"] = (
-    player_response
-      .dig?("playerOverlays", "playerOverlayRenderer", "endScreen", "watchNextEndScreenRenderer", "results")
-      .try &.as_a.compact_map { |r| parse_related r } || \
-       player_response
-        .dig?("webWatchNextResponseExtensionData", "relatedVideoArgs")
-        .try &.as_s.split(",").map { |r|
-          r = HTTP::Params.parse(r).to_h
-          JSON::Any.new(Hash.zip(r.keys, r.values.map { |v| JSON::Any.new(v) }))
-        }
-  ).try { |a| JSON::Any.new(a) } || JSON::Any.new([] of JSON::Any)
-
   # Top level elements
 
-  primary_results = player_response
-    .dig?("contents", "twoColumnWatchNextResults", "results", "results", "contents")
+  main_results = player_response.dig?("contents", "twoColumnWatchNextResults")
+
+  raise BrokenTubeException.new("twoColumnWatchNextResults") if !main_results
+
+  primary_results = main_results.dig?("results", "results", "contents")
+  secondary_results = main_results
+    .dig?("secondaryResults", "secondaryResults", "results")
+
+  raise BrokenTubeException.new("results") if !primary_results
+  raise BrokenTubeException.new("secondaryResults") if !secondary_results
 
   video_primary_renderer = primary_results
-    .try &.as_a.find(&.["videoPrimaryInfoRenderer"]?)
-      .try &.["videoPrimaryInfoRenderer"]
+    .as_a.find(&.["videoPrimaryInfoRenderer"]?)
+    .try &.["videoPrimaryInfoRenderer"]
```

Comment:
This is fine, but looking at all this, I'm wondering if a json path library would be useful for us https://github.com/json-path/JsonPath (this is a java one)

I've only ever used it in testing, but the equivalent of this would be an jsonpath of `$[?(@.videoPrimaryInfoRenderer)].videoPrimaryInfoRenderer` and it would return an array and we'd try to take the first item from it. Upside is that we'd have cleaner JSON searching, downside is that jsonpath is kind of confusing

Reply:
I don't think so.

The syntax is indeed quite confusing, and would make maintenance harder, especially for new comers (We already don't have a lot of contributions, I'd prefer to not require them to get familiar with such syntax).

The basic crstal syntax is good enough and quite legible, as long as we respect some styling rules to avoid a nightmare of `.try` statements and cascading conditions.